[glsl-in] fix missing side effects from sequence expressions

Vipitis opened the pull request:
**Connections**
fixes #6208

**Description**
Adds a `Sequence` expression kind so side effect of comma separated expressions aren't lost.

Comments welcome this is my first time trying to fix rust code.

**Testing**
Adds naga glsl test case, sanity checked against shadertoy, now both are visually grey. 

**Squash or Rebase?**
squash

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're not bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

Vipitis committed:
add test case

Vipitis committed:
add shader grammar annotations

Vipitis committed:
test comma expression in init part

Vipitis committed:
example "correct" translation

Vipitis committed:
handle expression sequences in the lowerer (works?)

Vipitis committed:
fix issue ref

Vipitis committed:
run formatting

Vipitis committed:
fix some  typos

Vipitis commented on naga/tests/in/glsl/multipart-for-loop.frag:
```diff
@@ -0,0 +1,19 @@
+// issue #6208 https://github.com/gfx-rs/wgpu/issues/6208
+# version 460
```
would it be preferred to adopt webgl cts tests, for example https://registry.khronos.org/webgl/sdk/tests/conformance/glsl/bugs/sequence-operator-evaluation-order.html?

inner-daemons replied:
Thats a question that @teoxoy will probably answer

teoxoy replied:
The GLSL frontend is GLSL 440+ with Vulkan semantics. WebGL uses GLSL ES 2/3 but as long as the semantics are the same we could use some of the WebGL CTS tests.

inner-daemons commented on naga/src/front/glsl/parser/expressions.rs:
```diff
@@ -511,20 +511,43 @@ impl ParsingContext<'_> {
+    // shader grammar :-
+    // expression :
+    // assignment_expression
+    // expression COMMA assignment_expression
```
Not sure this belongs here as a non-doc comment at the top of a very function

inner-daemons commented on naga/src/front/glsl/parser/expressions.rs:
```diff
@@ -511,20 +511,43 @@ impl ParsingContext<'_> {
+            Ok(expr)
+        } else {
+            let mut meta = stmt.hir_exprs[exprs[0]].meta;
+            for &e in &exprs[1..] {
+                meta.subsume(stmt.hir_exprs[e].meta);
+            }
```
Thanks for handling the spans correctly

inner-daemons commented on naga/src/front/glsl/context.rs:
```diff
@@ -1343,6 +1344,23 @@ impl<'a> Context<'a> {
+                        return Err(Error {
+                            kind: ErrorKind::SemanticError("Empty expression sequence".into()),
+                            meta,
+                        })
```
If this can't be produced by the parser its more clear to say `unreachable!()`

Vipitis replied:
I worried there might be cases where you end up with a sequence of just one element like `float a = ,0.1;` or `float b = (0.1,);` which could be troublesome - but those aren't allowed. will replace the block with the macro.

inner-daemons commented on naga/src/front/glsl/parser/functions.rs:
Lots of non-functional changes here, see overall review comment

inner-daemons reviewed (changes requested):
Thanks for this PR, its very useful I believe! One little nit in the comments plus one question:

Whats with all of these weirdly formatted comments? Are they taken from a spec somewhere? Just meant to guide you? I don't know if we should keep them or not

Vipitis commented:
I will address the review later (1-2 days hopefully). 

> Whats with all of these weirdly formatted comments? Are they taken from a spec somewhere? Just meant to guide you? I don't know if we should keep them or not

It's from the [GLSL spec](https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.60.pdf): chapter 9 shader grammar, to help me read the code initially. As the functions don't exactly match the productions and tokens there, I added them in the places that I looked at - to better understand what's happening. Sadly the grammar doesn't appear to be any kind of normalized CFG.
Similar comments already exist in `declarations.rs`, so would it be best to remove all, remove mine or add more (and agree on a format/prefix)?

inner-daemons commented:
@Vipitis With that in mind I'll do another pass on the comments. I definitely think many of them are misplaced at least.

Vipitis committed:
mark empty sequence unreachable

Vipitis committed:
remove shader grammar comments

Vipitis committed:
add changelog entry

inner-daemons commented on naga/src/front/glsl/parser/functions.rs:
```diff
@@ -432,9 +432,9 @@ impl ParsingContext<'_> {
                 meta
             }
+
             TokenValue::For => {
                 let mut meta = self.bump(frontend)?.meta;
-
                 ctx.symbol_table.push_scope();
```
If this change isn't necessary can you undo it?

inner-daemons reviewed (changes requested):
Looks good, nothing major! I think you've removed some of the spec related comments you had before, which wasn't necessary, but also won't be a problem. The only thing I'd ask you to do is look through the "files changed" tab and see which changes aren't necessary. This isn't important but I think we don't like having PRs change random line spacing for no good reason.

Vipitis committed:
revert white spaces

Vipitis commented:
I saw in the meeting notes it said waiting for author, and I wasn't 100% sure what there still was to do. So I removed the additional comments to make the decision easier and added a changelog note as that was part of the checklist (although the changelog in trunk already has two new subheaders for naga bug fixes - so to merge might not be clean).
Whitespaces should be reverted now.

I found one potential issue, will put it in a comment on that line.

Vipitis commented on naga/src/front/glsl/context.rs:
```diff
@@ -1343,6 +1344,18 @@ impl<'a> Context<'a> {
+                let mut last_handle = None;
+                for expr in exprs.iter() {
+                    let (handle, _) =
+                        self.lower_expect_inner(stmt, frontend, *expr, ExprPos::Rhs)?;
```
in the GLSL spec document it lists one change for 4.60, revision 4 [section 1.1.4](https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.60.pdf)

> * Private bug 13837: Ternary and sequence operators may operate on void types

the explanation between 4.50 and 4.60 (both times Section 5.9) are also slightly different. 
The later version adding two new phrases:

> ...  The operands to the sequence operator may have `void` type. Opaque
types cannot be used with the sequence (,) operator.

It is unclear to me if you target a specific version, or if the translation should care.

Perhaps it's  fine to use `lower_inner()` instead?

inner-daemons replied:
I'm a bit out of the loop on GLSL stuff. The parser is very much best-effort at this point. Its more about not breaking other people's code than implementing the GLSL spec in the most perfect way. So probably its your call what we should do here.

teoxoy replied:
I think it would be fine to use `lower_inner()` and accept void as a possible return type. The GLSL 4.4 spec also doesn't explicitly say that void is not allowed.

Vipitis replied:
I agree, and I can think of scenarios where this might be used. Although it feel like it shouldn't work on the right hand side of an assignment(the returned element). But when sequence expressions are used outside of an assignment like in loops it should be fine. I can't think of a solution on the spot.

WebGL2.0 specifically doesn't allow this which is all the user code I am working with.

inner-daemons commented:
Thanks for being so on top of this @Vipitis! Definitely it was a mistake to mark this as "waiting on author", I've just requested a review from @teoxoy.

teoxoy commented on naga/tests/in/glsl/multipart-for-loop.frag:
```diff
@@ -0,0 +1,19 @@
+
+    // a, b and c should be all ~0.5!
+    // right now it only ever takes the last expression in the block... 
+    // leading to infinite loops, lost devices and incorrect results.
```
```suggestion
```

This should be removed as this PR fixes this.

teoxoy commented on naga/src/front/glsl/ast.rs:
```diff
@@ -116,6 +116,10 @@ pub struct HirExpr {
 
 #[derive(Debug, Clone)]
 pub enum HirExprKind {
+    /// helper to represent a sequence of expressions where side effects matter, the last one is returned to the caller
```
```suggestion
    /// Represents a sequence of expressions. It returns the type and value of the last (i.e. right-most) expression.
```

The previous phrasing made it sound like this was related to function calls. I also removed the mention of side effects because we always use this variant for GLSL's sequence operator.

teoxoy reviewed (commented):
Looks good overall!

Vipitis committed:
Apply suggestions from code review

Co-authored-by: Teodor Tanasoaia <28601907+teoxoy@users.noreply.github.com>

teoxoy reviewed (approved):
Let's land this!
